testsuite/tests/typing-modules/package_constraint.ml

```diff
@@ -0,0 +1,171 @@
+(* You may constrain abstract types in packages. *)
+module type S = sig
+  type t
+end
+
+type m = (module S with type t = int);;
+[%%expect{|
+module type S = sig type t end
+type m = (module S with type t = int)
+|}];;
+
+(* You may use variables in the current environment in the new definitions. *)
+module type S = sig
+  type t
+end
+
+type 'a m = (module S with type t = 'a);;
+[%%expect{|
+module type S = sig type t end
+type 'a m = (module S with type t = 'a)
+|}];;
+
+(* It works with non-trivial paths. *)
+module type S = sig
+  module M : sig
+    type t
+  end
+end
+
+type m = (module S with type M.t = int)
+[%%expect{|
+module type S = sig module M : sig type t end end
+type m = (module S with type M.t = int)
+|}];;
+
+(* It should respect immediacy - [m1] should typecheck but not [m2]. *)
+module type S = sig
+  type t [@@immediate]
+end
+
+type m1 = (module S with type t = int)
+type m2 = (module S with type t = string);;
+[%%expect{|
+module type S = sig type t [@@immediate] end
+type m1 = (module S with type t = int)
+Line 6, characters 10-41:
+6 | type m2 = (module S with type t = string);;
+              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: In this "with" constraint, the new definition of "t"
+       does not match its original definition in the constrained signature:
+       Type declarations do not match:
+         type t = string
+       is not included in
+         type t [@@immediate]
+       The first is not an immediate type.
+|}];;
+
+(* You may not constrain types with a manifest in a package *)
+module type S = sig
+  type t = int
+end
+
+type m = (module S with type t = string);;
+[%%expect{|
+module type S = sig type t = int end
+Line 5, characters 9-40:
+5 | type m = (module S with type t = string);;
+             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: In the constrained signature, type "t" is defined to be "int".
+       Package "with" constraints may only be used on abstract types.
+|}];;
+
+(* Even if your constraint would be satisfied. *)
+(* It would be nice if this worked. *)
+module type S = sig
+  type t = int
+end
+
+type m = (module S with type t = int);;
+[%%expect{|
+module type S = sig type t = int end
+Line 5, characters 9-37:
+5 | type m = (module S with type t = int);;
+             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: In the constrained signature, type "t" is defined to be "int".
+       Package "with" constraints may only be used on abstract types.
+|}];;
+
+(* And even if the manifest is not obvious in the original definition. *)
+module M = struct
+  type t
+end
+
+module type S = sig
+  module P = M
+end
+
+type m = (module S with type P.t = int);;
+[%%expect{|
+module M : sig type t end
+module type S = sig module P = M end
+Line 9, characters 9-39:
+9 | type m = (module S with type P.t = int);;
+             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: In the constrained signature, type "P.t" is defined to be "M.t".
+       Package "with" constraints may only be used on abstract types.
+|}];;
+
```

Comment:
I think it may be useful in order to future proof to test the private row case:

```ocaml
module type Private_row = sig
  type a
  and t = private [< `A | `B ]
  and b
  and d = private [< `C ]
 end
 module type Test = Private_row with type t = [ `A ]
 type fail = (module Private_row with type t = [ `A ] )
 ```

Reply:
Good idea - test added.